sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala

```diff
@@ -375,15 +366,106 @@ case class GpuSum(child: Expression, resultType: DataType)
+ * required rearrangement of values into pivoted form.
+ *
+ * For example on an input of
+ * type | A | B
+ * -----+--+--
+ *   b | x | 1
+ *   a | x | 2
+ *   b | y | 3
+ *
+ * with type=groupbyKey, pivotColumn=A, valueColumn=B, and pivotColumnValues=[x,y]
+ *
+ * updateExpressions - In the partial_pivot stage, new columns are created based on
+ * pivotColumnValues one for each of the aggregation. Last aggregation on these columns grouped by
+ * `type` and convert into an array( as per Spark's expectation). Last aggregation(excluding nulls)
+ * works here as there would be atmost one entry in new columns when grouped by `type`.
+ * After CudfLastExcludeNulls, the intermediate result would be
+ *
+ * type | x | y
+ * -----+---+--
+ *   b | 1 | 3
+ *   a | 2 | null
+ *
+ *
+ * mergeExpressions - this is the final pivot aggregation after shuffle. We do another `Last`
+ * aggregation to merge the results. In this example all the data was combined in the
+ * partial_pivot hash aggregation. So it didn't do anything in this stage.
+ *
+ * The final result would be:
+ *
+ * type | x | y
+ * -----+---+--
+ *   b | 1 | 3
+ *   a | 2 | null
+ *
+ * @param pivotColumn column that determines which output position to put valueColumn in.
+ * @param valueColumn the column that is being rearranged.
+ * @param pivotColumnValues the list of pivotColumn values in the order of desired output. Values
+ *                          not listed here will be ignored.
+ */
+case class GpuPivotFirst(
+  pivotColumn: Expression,
+  valueColumn: Expression,
+  pivotColumnValues: Seq[Any]) extends GpuAggregateFunction {
+
+  val valueDataType = valueColumn.dataType
+
+  override val dataType: DataType = valueDataType
+  override val nullable: Boolean = false
+
+  val pivotColAttr = pivotColumnValues.map(pivotColumnValue => {
+    // If `pivotColumnValue` is null, then create an AttributeReference for null column.
+    if (pivotColumnValue == null) {
+      AttributeReference(GpuLiteral(null, valueDataType).toString, valueDataType)()
+    } else {
+      AttributeReference(pivotColumnValue.toString, valueDataType)()
+    }
+  })
+
+  override lazy val inputProjection: Seq[Expression] = {
+    val expr = pivotColumnValues.map(pivotColumnValue => {
+      if (pivotColumnValue == null) {
+        GpuIf(GpuIsNull(pivotColumn), valueColumn, GpuLiteral(null, valueDataType))
+      } else {
+        GpuIf(GpuEqualTo(pivotColumn, GpuLiteral(pivotColumnValue, pivotColumn.dataType)),
+          valueColumn, GpuLiteral(null, valueDataType))
+      }
+    })
+    expr
+  }
+
+  override lazy val updateExpressions: Seq[GpuExpression] = {
+    pivotColAttr.map(pivotColumnValue => new CudfLastExcludeNulls(pivotColumnValue))
+  }
+
+  override lazy val mergeExpressions: Seq[GpuExpression] = {
+    pivotColAttr.map(pivotColumnValue => new CudfLastExcludeNulls(pivotColumnValue))
+  }
+
+  override lazy val evaluateExpressions: Seq[Expression] = {
```

Comment:
So does this need to be a `Seq` still?

Reply:
I think it did at some point, but doesn't seem to anymore, and could remove some diffs.

Reply:
You are right. It's not required anymore.  I have removed it now.